File: 123_Best_Time_to_Buy_and_Sell_Stock_III/123_1.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        """
        a, b, c, d = 0, float("-inf"), 0, float("-inf")
        for p in prices:
            a = max(a, b + p)
            b = max(b, -p)
            c = max(c, d + p)
            d = max(d, a - p)
        return c
        """
        n = len(prices)
        dp = [[[0] * 2 for _ in range(3)] for _ in range(n)]
        for i in range(n):
            for k in range(2, 0, -1):
                if i - 1 == -1:
                    dp[i][k][0] = 0
                    dp[i][k][1] = -prices[i]
                    continue
                dp[i][k][0] = max(dp[i-1][k][0], dp[i-1][k][1] + prices[i])
                dp[i][k][1] = max(dp[i-1][k][1], dp[i-1][k-1][0] - prices[i])
        return dp[n-1][2][0]

File: 123_Best_Time_to_Buy_and_Sell_Stock_III/123_1.c
#include <stdio.h>
#include <stdlib.h>
#include <limits.h>
#include <stdbool.h>
#include <math.h>
#include <stdint.h>
#include <string.h>

int max(int a, int b) {
    return a > b ? a : b;
}

int maxProfit(int* prices, int pricesSize){
    if (!pricesSize) return 0;
    int dp[pricesSize][3][2];
    memset(dp, 0, sizeof(dp));
    dp[0][1][0] = 0;
    dp[0][1][1] = -prices[0];
    dp[0][2][0] = 0;
    dp[0][2][1] = -prices[0];
    for (int i = 1; i < pricesSize; i++) {
        dp[i][2][0] = max(dp[i - 1][2][0], dp[i - 1][2][1] + prices[i]);
        dp[i][2][1] = max(dp[i - 1][2][1], dp[i - 1][1][0] - prices[i]);
        dp[i][1][0] = max(dp[i - 1][1][0], dp[i - 1][1][1] + prices[i]);
        dp[i][1][1] = max(dp[i - 1][1][1], dp[i - 1][0][0] - prices[i]);
    }
    return dp[pricesSize - 1][2][0];
}



int main(void) {
    int prices[] = {7,1,5,3,6,4};
    printf("%d\n", maxProfit(prices, 6));
}

